### ISSUE TITLE
ipengine registers only locally?

### ISSUE THREAD

**Issue by dseuss** [ipython/ipython#8459]
I am trying to get the ipcluster to run on multiple machines using MPI. First, I register the controller using 

    > ipcontroller start --profile=mpi --ip='*'

which seems to work, since the hub listens on all interfaces (<code>tcp://*:52662</code>). The profile_mpi does not contain any nonstandard configuration. Also, the correct ip address is written to the <code>ipcontroller-engine.json</code>. However, when I run

    > ipengine start --profile=mpi --ip='*' --mpi=mpi4py

from inside a script launched with mpiexec, the engine on the same machine as the controller connects fine, but the other ones don't. The log looks like this

    [IPEngineApp] Registering with controller at tcp://127.0.0.1:54682
    [IPEngineApp] CRITICAL | Registration timed out after 30.0 seconds

(the increased timeout is the only configuration in the profile).

**Comment by minrk** [comment]
The only reason to specify the `--ip` arg on the engine is if the information in the connection file is *wrong*, and it should always be a real IP address.

**Comment by dseuss** [comment]
OK, even with that fixed, it still output
    
    Registering with controller at tcp://127.0.0.1:59179 <tcp://127.0.0.1:59179>

although the correct controller ip address is in the ipcontroller-engine.json file.

**Comment by minrk** [comment]
Are the controller and engine on the same machine? If so, 127.0.0.1 *is* the correct IP to connect to.

**Comment by dseuss** [comment]
Only partially. The ones that run on the same machine work fine, the ones on remote machines can’t connect.

**Comment by minrk** [comment]
@dseuss can you share the ipcontroller-engine.json file? The engine is determining that the controller is running on the same machine as the engine, apparently incorrectly. Typically this means they have an IP address in common.

Does it work if you pass `--ip=public_ip` instead of `--ip=*`?

**Comment by minrk** [comment]
@dseuss and the log output of the remote engines verifies that they are trying to connect to the correct IP?

**Comment by dseuss** [comment]
No success with <code>--ip=public_ip</code>. Here are the log and config files:

    $  cat .ipython/profile_mpi/security/ipcontroller-engine.json
    {
      "control": 33430,
      "task": 40887,
      "hb_ping": 35926,
      "mux": 50593,
      "pack": "json",
      "hb_pong": 44629,
      "ssh": "",
      "key": "ace4fb3f-5297-42b6-9682-4052c268dd22",
      "registration": 44932,
      "interface": "tcp://*",
      "iopub": 44723,
      "signature_scheme": "hmac-sha256",
      "unpack": "json",
      "location": "192.168.122.1"
    }

    $ cat ipengine-18925.log
    2015-05-23 10:15:53.500 [IPEngineApp] Initializing MPI:
    2015-05-23 10:15:53.500 [IPEngineApp] from mpi4py import MPI as mpi
    mpi.size = mpi.COMM_WORLD.Get_size()
    mpi.rank = mpi.COMM_WORLD.Get_rank()
    
    2015-05-23 10:15:53.501 [IPEngineApp] Loading url_file u'xxx/.ipython/profile_mpi/security/ipcontroller-engine.json'
    2015-05-23 10:15:53.524 [IPEngineApp] Registering with controller at tcp://127.0.0.1:44932
    2015-05-23 10:16:23.553 [IPEngineApp] CRITICAL | Registration timed out after 30.0 seconds

    $ cat ipengine-19274.log
    2015-05-23 10:15:53.144 [IPEngineApp] Initializing MPI:
    2015-05-23 10:15:53.145 [IPEngineApp] from mpi4py import MPI as mpi
    mpi.size = mpi.COMM_WORLD.Get_size()
    mpi.rank = mpi.COMM_WORLD.Get_rank()

    2015-05-23 10:15:53.145 [IPEngineApp] Loading url_file u'xxx/.ipython/profile_mpi/security/ipcontroller-engine.json'
    2015-05-23 10:15:53.155 [IPEngineApp] Registering with controller at tcp://127.0.0.1:44932
    2015-05-23 10:15:53.220 [IPEngineApp] Starting to monitor the heartbeat signal from the hub every 3010 ms.
   
    2015-05-23 10:15:53.994 [IPEngineApp] Completed registration with id 0

**Comment by minrk** [comment]
Can you show the connection files and logs when you use `--ip=public_ip` as well?

Can you also share the output of `ip -f inet addr | grep inet` on the engine machines, both remote and local?

**Comment by dseuss** [comment]
Sorry for the long delay, but I wasn't in the office last week. Here is the complete output for <code>--ip=public_ip</code> and <code>ip -f inet addr</code>.

    > ip -f inet addr | grep inet
    # Node that runs ipcluster
    inet 127.0.0.1/8 scope host lo
    inet 10.16.1.142/16 scope global eth0
    inet 10.12.1.142/16 brd 10.12.255.255 scope global ib0
    inet 192.168.122.1/24 brd 192.168.122.255 scope global virbr0

    # Remote node
    inet 127.0.0.1/8 scope host lo
    inet 10.16.1.141/16 scope global eth0
    inet 10.12.1.141/16 brd 10.12.255.255 scope global ib0
    inet 192.168.122.1/24 brd 192.168.122.255 scope global virbr0

    > cat .ipython/profile_mpi/log/ipcontroller-1549.log
    2015-06-03 19:27:54.964 [IPControllerApp] Hub listening on tcp://*:45756 for registration.
    2015-06-03 19:27:54.966 [IPControllerApp] Hub using DB backend: 'NoDB'
    2015-06-03 19:27:55.230 [IPControllerApp] hub::created hub
    2015-06-03 19:27:55.230 [IPControllerApp] writing connection info to xxx.ipython/profile_mpi/security/ipcontroller-client.json
    2015-06-03 19:27:55.232 [IPControllerApp] writing connection info to xxx.ipython/profile_mpi/security/ipcontroller-engine.json
    2015-06-03 19:27:55.233 [IPControllerApp] task::using Python leastload Task scheduler
    2015-06-03 19:27:55.233 [IPControllerApp] Heartmonitor started
    2015-06-03 19:27:55.238 [IPControllerApp] Creating pid file: xxx.ipython/profile_mpi/pid/ipcontroller.pid
    2015-06-03 19:27:55.255 [IPControllerApp] client::client '\x00\xe98@\xa8' requested u'connection_request'
    2015-06-03 19:27:55.255 [IPControllerApp] client::client ['\x00\xe98@\xa8'] connected
    2015-06-03 19:28:35.480 [IPControllerApp] client::client '0de21428-8a22-416b-8492-7e32db6ec40d' requested u'registration_request'
    2015-06-03 19:28:40.236 [IPControllerApp] registration::finished registering engine 0:0de21428-8a22-416b-8492-7e32db6ec40d
    2015-06-03 19:28:40.237 [IPControllerApp] engine::Engine Connected: 0

    > cat .ipython/profile_mpi/log/ipengine-1587.log
    2015-06-03 19:28:35.467 [IPEngineApp] Initializing MPI:
    2015-06-03 19:28:35.468 [IPEngineApp] from mpi4py import MPI as mpi
    mpi.size = mpi.COMM_WORLD.Get_size()
    mpi.rank = mpi.COMM_WORLD.Get_rank()

    2015-06-03 19:28:35.469 [IPEngineApp] Loading url_file u'xxx.ipython/profile_mpi/security/ipcontroller-engine.json'
    2015-06-03 19:28:35.478 [IPEngineApp] Registering with controller at tcp://127.0.0.1:45756
    2015-06-03 19:28:35.547 [IPEngineApp] Starting to monitor the heartbeat signal from the hub every 3010 ms.
    2015-06-03 19:28:35.551 [IPEngineApp] Running file in user namespace: xxx.ipython/profile_default/startup/10-magic.ipy
    2015-06-03 19:28:35.870 [IPEngineApp] Running file in user namespace: xxx.ipython/profile_default/startup/20-imports.py
    2015-06-03 19:28:36.395 [IPEngineApp] Completed registration with id 0

    > cat .ipython/profile_mpi/log/ipengine-1966.log
    2015-06-03 19:28:35.639 [IPEngineApp] Initializing MPI:
    2015-06-03 19:28:35.639 [IPEngineApp] from mpi4py import MPI as mpi
    mpi.size = mpi.COMM_WORLD.Get_size()
    mpi.rank = mpi.COMM_WORLD.Get_rank()

    2015-06-03 19:28:35.641 [IPEngineApp] Loading url_file u'xxx.ipython/profile_mpi/security/ipcontroller-engine.json'
    2015-06-03 19:28:35.651 [IPEngineApp] Registering with controller at tcp://127.0.0.1:45756
    2015-06-03 19:29:05.682 [IPEngineApp] CRITICAL | Registration timed out after 30.0 seconds

    > cat .ipython/profile_mpi/security/ipcontroller-engine.json
    {
      "control": 46244,
      "task": 47408,
      "hb_ping": 51534,
      "mux": 57636,
      "pack": "json",
      "hb_pong": 52690,
      "ssh": "",
      "key": "77aeede5-0d4b-4e8e-8660-0b7f78d85d60",
      "registration": 45756,
      "interface": "tcp://*",
      "iopub": 45561,
      "signature_scheme": "hmac-sha256",
      "unpack": "json",
      "location": "192.168.122.1"
    }

It's a remotely operated cluster, so I can't tell you more on its connectivity graph, but to me it looks like the ipcluster/ipengines are using the <code>vibr0</code> interface where they should be using the <code>eth0</code> interface.

**Comment by rootbe** [comment]
Exactly,  your controller is sending 192.168.121.1 as the IP that engines should use to connect to it. This won't work as both machines use this IP address on virbr0.
Try setting one of the 10.x.x.x addresses your controller uses explicitly via the location parameter and this should be fixed.
